operator/pkg/client/client_factory.go

```diff
@@ -83,7 +87,8 @@ var (
 	_ action.RESTClientGetter = &SingletonClients{}
 )
 
-func NewSingletonClients(config *rest.Config, logger logr.Logger) (*SingletonClients, error) {
+func NewSingletonClients(config *rest.Config, logger logr.Logger, existingRuntimeClient client.Client,
```

Comment:
IMO instead of passing in a client here, we should change all occurrences to use the client from singletonClients instead. This way you can make sure that there are never more clients initialized. IMO you should initialize the singleton clients always and then cache them immediately and then right after that always use the runtime client.

Reply:
That would not suffice the single cluster use case, where the idea is to re-use to source runtime client. I will think of a way..

Reply:
I am suggesting you to create the singleton client first and then pass it down. The order of creation should not matter and the cache has to be build up the same way.

Reply:
I know what you mean, would have looked better hadn't it been the library use case. The Manifest preparation logic is only intended to translate its spec to a human readable format and call the library functions, so SingletonClient doesn't belong to the pre-processing part - unless we change the architecture